algorithms/mopt/mopt_hierarflow_routines.cpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2022 Ricerca Security
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+#include "fuzzuf/algorithms/mopt/mopt_hierarflow_routines.hpp"
+
+#include "fuzzuf/algorithms/mopt/mopt_optimizer.hpp"
+#include "fuzzuf/algorithms/mopt/mopt_option.hpp"
+#include "fuzzuf/algorithms/mopt/mopt_option_get_splice_cycles.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+namespace fuzzuf::algorithm::mopt::routine {
+
+namespace other {
+
+using mopt::option::MOptTag;
+
+MOptUpdate::MOptUpdate(MOptState &state) : state(state) {}
+
+MOptMidCalleeRef MOptUpdate::operator()(
+    [[maybe_unused]] std::shared_ptr<MOptTestcase> testcase) {
+  // retrive optimizer values
+  auto last_splice_cycle = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::LastSpliceCycle, true);
+  auto havoc_operator_finds = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::HavocOperatorFinds, true);
+
+  auto &mopt = state.mopt;
+
+  // update havoc_operator_finds
+  for (size_t i = 0; i < havoc_operator_finds.size(); i++) {
+    mopt->havoc_operator_finds[state.core_mode ? 1 : 0][i] +=
+        havoc_operator_finds[i];
+  }
+
+  if (last_splice_cycle >= afl::option::GetSpliceCycles(state)) {
+    state.UpdateSpliceCycles();
+  }
+
+  if (state.pacemaker_mode) {
+    // TODO
+  }
+
+  auto new_testcases = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::NewTestcases, true);
+  auto selected_case_histogram = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::SelectedCaseHistogram, true);
+
+  // pilot mode (update local best)
+  if (!state.core_mode) {
+    if (unlikely(new_testcases > mopt::option::GetPeriodPilot<MOptTag>())) {
+      for (size_t i = 0; i < selected_case_histogram.size(); i++) {
+        double score = 0.0;
+        if (selected_case_histogram[i] > 0) {
+          score = (mopt->havoc_operator_finds[0][i] +
+                   mopt->havoc_operator_finds[1][i]) /
+                  selected_case_histogram[i];
+        }
+        mopt->SetScore(i, score);
+      }
+      mopt->UpdateLocalBest();
+
```

Comment:
After (trying) the update of local best, you need to reset `havoc_operator_finds` of pilot mode and the corresponding `(accm_)selected_case_histogram` .

Reply:
fixed in b386bc7a39218cd286ab99b8e2fc51af97ead618

Reply:
Oh wait. There is another issue.

You don't check if the new score is higher than the one recorded in the optimizer.

Compare the followings:
https://github.com/puppet-meteor/MOpt-AFL/blob/master/MOpt/afl-fuzz.c#L8717-L8721

and

https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-77ad669e643808933880a044def0763a1788478a9f9a42bea615b130fdb42438R83
https://github.com/fuzzuf/fuzzuf/pull/84/files#diff-6a8c4972fbfb3703282f4f41c9b8a685401f29bc8c395d701a71bb15fdf0d496R86-R89

Reply:
`UpdateLocalBest` will check if the score is higher or not, which is called after all scores set
So I think this would not be an issue.

Reply:
Ah that's right 
sorry don't care